### PR TITLE
lambda - Fix flaky integration tests which assume there are no other functions in the account

### LINKED ISSUE #1277: lambda integration tests - potential test flake
Another issue I just found: **[lambda]** Integration tests fail if the account is not empty. The lambda integration tests assume there are no other lambdas in the account. I just swapped my region to one I don't use. 

The error was for integration testing lambda. It was an assertion error.

```
  - name: lambda_info | Ensure default query value is 'config' when function name
      omitted
    lambda_info:
    register: lambda_infos_query_config
    check_mode: yes
  - name: lambda_info | Assert successfull retrieval of all information
    assert:
      that:
...
      - lambda_infos_query_config.functions[0].function_name == lambda_function_name
      - 
```

If the freshly created lambda for this test is not the first lambda in the list (because a pre-existing lambda is), then this assertion fails.

Perhaps it's just that this specific test needs to be made a bit smarter.

_Originally posted by @mdavis-xyz in https://github.com/ansible-collections/amazon.aws/issues/924#issuecomment-1319708366_

### Issue Type

Bug Report

### Component Name

lambda

### Ansible Version

CI / latest

### Collection Versions

CI / latest

### Configuration

CI

### OS / Environment

CI

### Steps to Reproduce

Run lambda integration tests in an account with existing lambda

### Expected Results

Run without errors

### Actual Results

Assertion failure

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by mdavis-xyz**
##### SUMMARY

The integration tests for lambda assume that there are no other lambdas in the account.
So when we index into `[0]`, that's not necessarily the lambda we just created.

Note that this fix doesn't work. One of the assertions is failing.

```
TASK [lambda : lambda_info | Assert successfull retrieval of all information 2] ***
fatal: [testhost]: FAILED! => {
    "assertion": "lambda_infos_all.function[lambda_function_name].versions is not defined",
    "changed": false,
    "evaluated_to": false,
    "msg": "Assertion failed"
}
```

That attribute is defined. It's a list of 2 versions. I don't know why.
I'm just creating this PR anyway, so someone else can pickup where I left off.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
lambda_info

##### ADDITIONAL INFORMATION

Fixes #1277

### CODE REVIEW

File: tests/integration/targets/lambda/tasks/main.yml [hunk 1]
```diff
@@ -315,49 +315,55 @@
...
+      - lambda_function_name in lambda_infos_query_config.function
+      - lambda_infos_all.function[lambda_function_name].runtime == lambda_python_runtime
+      - lambda_infos_all.function[lambda_function_name].description == ""
+      - lambda_infos_all.function[lambda_function_name].function_arn is defined
+      - lambda_infos_all.function[lambda_function_name].handler == lambda_python_handler
+      - lambda_infos_all.function[lambda_function_name].versions is not defined
+      - lambda_infos_all.function[lambda_function_name].aliases is not defined
+      - lambda_infos_all.function[lambda_function_name].policy is not defined
+      - lambda_infos_all.function[lambda_function_name].mappings is not defined
+      - lambda_infos_all.function[lambda_function_name].tags is not defined
```
Comment (tremble) [inline tests/integration/targets/lambda/tasks/main.yml:360]:
```suggestion
      - lambda_function_name in lambda_infos_query_config.function
      - lambda_infos_query_config.function[lambda_function_name].runtime == lambda_python_runtime
      - lambda_infos_query_config.function[lambda_function_name].description == ""
      - lambda_infos_query_config.function[lambda_function_name].function_arn is defined
      - lambda_infos_query_config.function[lambda_function_name].handler == lambda_python_handler
      - lambda_infos_query_config.function[lambda_function_name].versions is not defined
      - lambda_infos_query_config.function[lambda_function_name].aliases is not defined
      - lambda_infos_query_config.function[lambda_function_name].policy is not defined
      - lambda_infos_query_config.function[lambda_function_name].mappings is not defined
      - lambda_infos_query_config.function[lambda_function_name].tags is not defined
```

**Comment by tremble** [comment]
Issue was introduced by https://github.com/ansible-collections/community.aws/pull/1239 - the dict of dicts was deprecated for consistency with other modules.

We probably need to do something clever with selectattr :/

File: tests/integration/targets/lambda/tasks/main.yml [hunk 2]
```diff
@@ -315,49 +315,55 @@
...
     register: lambda_infos_query_config
     check_mode: yes
-  - name: lambda_info | Assert successfull retrieval of all information
+  - name: debug
+    debug:
+      var: lambda_infos_all
+  - name: debug function name
+    debug:
+      var: lambda_function_name
+  - name: lambda_info | Assert successfull retrieval of all information 2
```
[the remark this answers is not in the document]
Comment (mdavis-xyz) [reply tests/integration/targets/lambda/tasks/main.yml]: I've removed the debug statements completely. I didn't mean to commit that.

**Comment by mdavis-xyz** [comment]
I've changed the tests to use `selectattr` now, instead of the deprecated `.function`.

But the assert `.versions is not defined` test fails. Why?

File: tests/integration/targets/lambda/tasks/main.yml [hunk 3]
```diff
@@ -315,49 +315,49 @@
...
+      - (lambda_infos_all.functions | selectattr('function_name', 'eq', lambda_function_name) | first).description == ""
+      - (lambda_infos_all.functions | selectattr('function_name', 'eq', lambda_function_name) | first).function_arn is defined
+      - (lambda_infos_all.functions | selectattr('function_name', 'eq', lambda_function_name) | first).handler == lambda_python_handler
+      - (lambda_infos_all.functions | selectattr('function_name', 'eq', lambda_function_name) | first).versions is defined
+      - (lambda_infos_all.functions | selectattr('function_name', 'eq', lambda_function_name) | first).aliases is defined
+      - (lambda_infos_all.functions | selectattr('function_name', 'eq', lambda_function_name) | first).policy is defined
+      - (lambda_infos_all.functions | selectattr('function_name', 'eq', lambda_function_name) | first).mappings is defined
+      - (lambda_infos_all.functions | selectattr('function_name', 'eq', lambda_function_name) | first).tags is defined
+      - (lambda_infos_all.functions | selectattr('function_name', 'eq', lambda_function_name) | first).architectures == ['arm64']
```
Comment (tremble) [inline tests/integration/targets/lambda/tasks/main.yml:333]:
```suggestion
      - tmp_function_info.architectures == ['arm64']
    vars:
      tmp_function_info: lambda_infos_all.functions | selectattr('function_name', 'eq', lambda_function_name) | first
```

When the link starts getting this long I sometimes use a temporary variable to try and make things easier to follow

File: tests/integration/targets/lambda/tasks/main.yml [hunk 4]
```diff
@@ -315,49 +315,49 @@
...
+      - (lambda_infos_all.functions | selectattr('function_name', 'eq', lambda_function_name) | first).runtime == lambda_python_runtime
+      - (lambda_infos_all.functions | selectattr('function_name', 'eq', lambda_function_name) | first).description == ""
+      - (lambda_infos_all.functions | selectattr('function_name', 'eq', lambda_function_name) | first).function_arn is defined
+      - (lambda_infos_all.functions | selectattr('function_name', 'eq', lambda_function_name) | first).handler == lambda_python_handler
+      - (lambda_infos_all.functions | selectattr('function_name', 'eq', lambda_function_name) | first).versions is not defined
+      - (lambda_infos_all.functions | selectattr('function_name', 'eq', lambda_function_name) | first).aliases is not defined
+      - (lambda_infos_all.functions | selectattr('function_name', 'eq', lambda_function_name) | first).policy is not defined
+      - (lambda_infos_all.functions | selectattr('function_name', 'eq', lambda_function_name) | first).mappings is not defined
+      - (lambda_infos_all.functions | selectattr('function_name', 'eq', lambda_function_name) | first).tags is not defined
```
Comment (tremble) [inline tests/integration/targets/lambda/tasks/main.yml:354]: This block should be testing `lambda_infos_query_config` rather than `lambda_infos_all`

I also find that `in` behaves more consistently than `defined`:  `- '"policy" not in ...'` or `- '"function_arn" in ...'` There are some gotchas around exactly what counts as "defined" when working in Ansible/Jinja2

**Review by tremble — CHANGES_REQUESTED** [review]
I think the "defined" issue you're seeing is caused by testing against `lambda_infos_all` rather than `lambda_infos_query_config`

I have a slight preference for using a temporary variable when using something so complex to access a list/dict item, but changing that's certainly not a blocker.

**Review by tremble — APPROVED** [review]

**Comment by tremble** [comment]
Thanks for the fix.
